jgiven-core/src/main/java/com/tngtech/jgiven/impl/intercept/ByteBuddyMethodInterceptor.java

```diff
@@ -17,54 +18,71 @@
+        if (interceptor == null) {
             return zuper.call();
         }
 
-        return interceptor.intercept( receiver, method, parameters, zuper::call );
+        return interceptor.intercept(receiver, method, parameters, zuper::call);
     }
 
     @RuntimeType
-    @BindingPriority( BindingPriority.DEFAULT * 2 )
-    public Object interceptDefault( @DefaultCall final Callable<?> zuper, @This final Object receiver, @Origin Method method,
-            @AllArguments final Object[] parameters,
-            @FieldProxy( INTERCEPTOR_FIELD_NAME ) StepInterceptorGetterSetter stepInterceptorGetter )
-            throws Throwable{
+    @BindingPriority(BindingPriority.DEFAULT * 2)
+    public Object interceptDefault(@DefaultCall final Callable<?> zuper, @This final Object receiver, @Origin Method method,
+                                   @AllArguments final Object[] parameters,
+                                   @FieldProxy(INTERCEPTOR_FIELD_NAME) StepInterceptorGetterSetter stepInterceptorGetter)
+            throws Throwable {
 
         StepInterceptor interceptor = (StepInterceptor) stepInterceptorGetter.getValue();
+        updateStageNameWrapper(receiver, method);
 
-        if( interceptor == null ) {
+        if (interceptor == null) {
             return zuper.call();
         }
 
-        return interceptor.intercept( receiver, method, parameters, zuper::call );
+        return interceptor.intercept(receiver, method, parameters, zuper::call);
     }
 
     @RuntimeType
-    public Object intercept( @This final Object receiver, @Origin final Method method,
-            @AllArguments final Object[] parameters,
-            @FieldProxy( INTERCEPTOR_FIELD_NAME ) StepInterceptorGetterSetter stepInterceptorGetter )
-            throws Throwable{
+    public Object intercept(@This final Object receiver, @Origin final Method method,
+                            @AllArguments final Object[] parameters,
+                            @FieldProxy(INTERCEPTOR_FIELD_NAME) StepInterceptorGetterSetter stepInterceptorGetter)
+            throws Throwable {
         // this intercepted method does not have a non-abstract super method
         StepInterceptor interceptor = (StepInterceptor) stepInterceptorGetter.getValue();
+        updateStageNameWrapper(receiver, method);
 
-        if( interceptor == null ) {
+        if (interceptor == null) {
             return null;
         }
 
-        return interceptor.intercept( receiver, method, parameters, () -> null );
+        return interceptor.intercept(receiver, method, parameters, () -> null);
+    }
+
+    private void updateStageNameWrapper(Object stageObject, Method method) {
+        StageNameWrapper methodNameWrapper = getMethodNameWrapper(method);
+        if (methodNameWrapper != null) {
+            ((StageNameInternal) stageObject).__jgiven_setStageNameWrapper(methodNameWrapper);
+        }
+    }
+
+    private StageNameWrapper getMethodNameWrapper(Method method) {
```

Comment:
using null is not as bad in private methods, but still I'd like to know if `Optional<StageNameWrapper>` would wokr here.

Reply:
I think it would make things harder if we are implementing it with an Optional. If it is set to null, we deal with it in the methods from ScenarioExecutor and we are also using Objects.equals as you suggested.